File: Java/Problem14.java
import java.io.*;
import java.util.*;
import java.time.*;

class Problem14 {
    public static void main(String args[]){
        try{
            Instant i0 = Instant.now();
            File file = new File("/home/ec2-user/environment/AOC/Resources/problem14.txt");
            
            System.out.println("Part 1: "+part1(file));
            Instant i1 = Instant.now();
            System.out.println("Part 2: "+part2(file));
            Instant i2 = Instant.now();
            System.out.println(String.format("Part 1: %s, Part 2: %s", 
                Duration.between(i0, i1).toMillis(),
                Duration.between(i1, i2).toMillis()));
            
        } catch (Exception e) {
            e.printStackTrace();
        }
        
    }
    
    public static long part1(File file) throws Exception{
        BufferedReader br = new BufferedReader(new FileReader(file));
        Mask m = new Mask("");
        String current;
        HashMap<Integer,Long> memory = new HashMap<Integer,Long>();
        while ((current = (br.readLine())) != null) {
            String[] split = current.split(" = ");
            if(split[0].equals("mask")){
                m = new Mask(split[1]);
            } else {
                Integer locInt = Integer.parseInt(split[0].substring(4, split[0].indexOf("]")));
                memory.put(locInt, m.apply(Integer.parseInt(split[1])));
            }
        }
        
        long returnVal = 0L;
        for(Long val : memory.values()){
            returnVal += val;
        }
        return returnVal;
    }
    
    public static long part2(File file) throws Exception{
        BufferedReader br = new BufferedReader(new FileReader(file));
        AddressMask m = new AddressMask("");
        String current;
        HashMap<Long,Long> memory = new HashMap<Long,Long>();
        ArrayList<Long> memoryLocs;
        while ((current = (br.readLine())) != null) {
            String[] split = current.split(" = ");
            if(split[0].equals("mask")){
                m = new AddressMask(split[1]);
            } else {
                Long loc = Long.parseLong(split[0].substring(4, split[0].indexOf("]")));
                Long value = Long.parseLong(split[1]);
                memoryLocs = m.apply(loc);
                for(Long l : memoryLocs){
                    memory.put(l, value);
                }
            }
        }
        
        long returnVal = 0L;
        for(Long val : memory.values()){
            returnVal += val;
        }
        return returnVal;
    }
}

class Mask{
    long orMask;
    long andMaskInverse;
    public Mask(String s){
        orMask = 0;
        andMaskInverse = 0;
        
        long twos = 1L;
        for(int i = s.length()-1; i >= 0; i--){
            if(s.charAt(i) == '0'){
                andMaskInverse += twos;
            } else if(s.charAt(i) == '1'){
                orMask += twos;
            }
            twos *= 2;
        }
    }
    
    public long apply(int num){
        return (num & ~andMaskInverse) | orMask;
    }
}

class AddressMask{
    long orMask;
    ArrayList<Long> floatingTwos;
    public AddressMask(String s){
        orMask = 0;
        floatingTwos = new ArrayList<Long>();
        
        long twos = 1L;
        for(int i = s.length()-1; i >= 0; i--){
            if(s.charAt(i) == '1'){
                orMask += twos;
            } else if(s.charAt(i) == 'X'){
                floatingTwos.add(twos);
            }
            twos *= 2;
        }
    }
    
    //returns list of all valid locations
    public ArrayList<Long> apply(long num){
        num = num | orMask;
        ArrayList<Long> toReturn = new ArrayList<Long>();
        toReturn.add(num);
        ArrayList<Long> temp = new ArrayList<Long>();
        
        for(Long two : floatingTwos){
            for(Long n : toReturn){
                temp.add(n | two);
                temp.add(n & ~(two));
            }
            toReturn.clear();
            for(Long x : temp){
                toReturn.add(x);
            }
            temp.clear();
        }
        
        return toReturn;
    }
}

File: 2020/Python/problem14.py
def part1():
    memory = dict()
    andMaskInverse = orMask = ""
    for line in open("../Resources/problem14.txt"):
        raw = line.split(" = ")
        if line[0:4] == "mask":
            andMaskInverse = orMask = ""
            for c in raw[1].strip():
                andMaskInverse = andMaskInverse + ("1" if c == "0" else "0")
                orMask = orMask + ("1" if c == "1" else "0")
        else:
            value = int(raw[1])
            masked_value = (value & ~int(andMaskInverse, 2)) | int(orMask, 2)
            address = int(raw[0][4:-1])
            memory[address] = masked_value
    part1 = 0
    for k in memory:
        part1 += memory[k]
    return part1

def part2():
    memory = dict()
    orMask = ""
    floatingTwos = []
    for line in open("../Resources/problem14.txt"):
        raw = line.split(" = ")
        if line[0:4] == "mask":
            orMask = ""
            floatingTwos = []
            currentTwo = 1
            raw_mask = raw[1].strip()
            for i in range(len(raw_mask)-1,-1,-1):
                if raw_mask[i] == "X":
                    floatingTwos.append(currentTwo)
                orMask = ("1" if raw_mask[i] == "1" else "0") + orMask
                currentTwo = currentTwo * 2
        else:
            value = int(raw[1])
            address = int(raw[0][4:-1]) | int(orMask, 2)
            end_addresses = [address]
            temp = []
            for two in floatingTwos:
                for a in end_addresses:
                    temp.append(a | two)
                    temp.append(a & ~two)
                end_addresses.clear()
                for a in temp:
                    end_addresses.append(a)
                temp.clear()
            for a in end_addresses:
                memory[a] = value
    part2 = 0
    for k in memory:
        part2 += memory[k]
    return part2

print("Part 1:",part1())
print("Part 2:",part2())